UTF-16 support

jfkthame opened the pull request:
This aims to provide "native" UTF-16 support, with `*U16` versions of `InitialInfo`, `BidiInfo`, and `Paragraph` structs that are parallel to the existing types but work on a `[u16]` text buffer instead of a `str`.

Fixes https://github.com/servo/unicode-bidi/issues/108

Manishearth commented on src/lib.rs:
```diff
@@ -829,11 +861,282 @@ fn assign_levels_to_removed_chars(para_level: Level, classes: &[BidiClass], leve
+        }
+        // Get the indicated code unit; if it's not a surrogate, we're done.
+        let c = self[index];
+        if !is_surrogate(c) {
```
nit: please just use the checked version here and don't check for surrogateyness

jfkthame replied:
Done (in following nit-fix commit)

Manishearth commented:
This is pretty big and I'm unlikely to be able to review it soon, fwiw.

jfkthame commented on src/lib.rs:
```diff
@@ -829,11 +861,282 @@ fn assign_levels_to_removed_chars(para_level: Level, classes: &[BidiClass], leve
+        // Otherwise, try to decode, returning REPLACEMENT_CHARACTER for errors.
+        // Note that the iterator returned by char::decode_utf16 cannot be empty
+        // because we already checked that we're not at the end of the text;
+        // therefore, the call to .next() will not return None.
```
Done (in following nit-fix commit)

jfkthame commented:
> This is pretty big and I'm unlikely to be able to review it soon, fwiw.

Yeah - sorry, I know there's a lot of it!

To make it more approachable, I tried to post it as a series of logical steps that build on each other (and I think it builds and passes tests at all stages of the patch queue, IIRC).

Much of it is moving (rather than modifying) existing code, and then duplicating (with `str` -> `[u16]` changes) the structs that are the public API. So there's actually not much new code here; once the TextSource trait is available, the rest is pretty mechanical.

Manishearth commented on src/lib.rs:
```diff
@@ -829,11 +860,282 @@ fn assign_levels_to_removed_chars(para_level: Level, classes: &[BidiClass], leve
+            if let Ok(ch) = ch {
+                // This must be a surrogate pair, otherwise char::from_u32() above should
+                // have succeeded!
+                assert!(ch.len_utf16() == 2);
```
nit: debug_assert (here and below)

We should avoid panicking in this crate.

Manishearth commented on src/lib.rs:
```diff
@@ -91,15 +91,46 @@ pub use crate::prepare::LevelRun;
+
+    /// An iterator over the text returning Unicode characters,
+    /// REPLACEMENT_CHAR for invalid code units.
+    fn chars(&self) -> Box<dyn DoubleEndedIterator<Item = char> + '_>;
```
issue: please no allocation or dynamic dispatch for this kind of thing


Instead, please add associated types for iterators. Eventually we can use type-alias-impl-trait.

Also I don't want to commit to this API just yet. I suggest is making this a sealed trait with `doc(hidden)` methods, and documenting it as unstable for implementation or direct use.

Manishearth commented on src/implicit.rs:
```diff
@@ -120,9 +120,9 @@ pub fn resolve_weak(
                     // We want to make sure we check the correct next character by skipping past the rest
                     // of this one.
-                    if let Some(ch) = text.get(i..).and_then(|s| s.chars().next()) {
+                    if let Some((_, len)) = text.char_at(i) {
```
nit: call it `char_len`

Manishearth commented on src/implicit.rs:
```diff
@@ -288,12 +288,12 @@ pub fn resolve_neutral<D: BidiDataSource>(
         let mut class_to_set = None;
 
-        let start_len_utf8 = text[pair.start..].chars().next().unwrap().len_utf8();
+        let (_, start_len) = text.char_at(pair.start).unwrap();
```
nit: here and elsewhere, I'd call these things `start_char_len` (etc)

Manishearth commented on src/implicit.rs:
```diff
@@ -487,27 +487,17 @@ fn identify_bracket_pairs<D: BidiDataSource>(
+        let mut actual_index = level_run.start;
+
+        while actual_index < level_run.end {
+            let (ch, len) = text.char_at(actual_index).unwrap();
```
issue: no panics, please.

The older structure of the code was intentional to use `char_indices`, please retain something similar. I think we can add an indexing method that returns a `&Self` to `TextSource` and then call `char_indices()` on it

Manishearth commented on src/lib.rs:
```diff
@@ -239,120 +238,132 @@ impl<'text> InitialInfoExt<'text> {
+}
+
+/// Implementation of initial-info computation.
+fn compute_initial_info<D: BidiDataSource, T: TextSource + ?Sized>(
```
assuming that the diffs to this function in commit 3df9426863f72482fd7a4ee36e44692c6ad6a084 are idempotent. looks like some match arms were moved around leading to a larger diff than this actually is, and I don't see any semantic changes.

Please let me know if my read here is incorrect!

Manishearth reviewed (commented):
Review has reached e3bad1f8474a7cdfa68dc9348ed1c7c1428560bf, which seems rather big and might take a while to review

Manishearth commented on src/lib.rs:
```diff
@@ -739,76 +597,260 @@ impl<'text> BidiInfo<'text> {
+    assign_levels_to_removed_chars(para.level, original_classes, levels);
+}
+
+fn reorder_levels<T: TextSource + ?Sized>(
```
nit: please also document these functions and what range of spec rules they follow

Manishearth commented on src/lib.rs:
```diff
@@ -709,6 +709,259 @@ impl<'text> BidiInfo<'text> {
+/// vectors.
+// TODO: Impl `struct StringProperty<T> { values: Vec<T> }` and use instead of Vec<T>
+#[derive(Debug, PartialEq)]
+pub struct BidiInfoU16<'text> {
```
thought: from a public API point of view, I wonder if this should all go in a `utf16` module.

From a code organization POV it probably should.

Manishearth commented on src/lib.rs:
```diff
@@ -709,6 +709,259 @@ impl<'text> BidiInfo<'text> {
+    /// [Rule L4]: https://www.unicode.org/reports/tr9/#L4
+    #[cfg_attr(feature = "flame_it", flamer::flame)]
+    pub fn reorder_line(&self, para: &ParagraphInfo, line: Range<usize>) -> Cow<'text, [u16]> {
+        if !level::has_rtl(&self.levels[line.clone()]) {
```
suggestion: would be nice to also refactor this out, perhaps in a way that allows the caller to pass in the `encode_utf16`

jfkthame replied:
I didn't do this for now, as it's a pretty short function and it wasn't really clear to me how to do it most tidily. But maybe as a followup...

Manishearth commented on src/lib.rs:
```diff
@@ -1590,6 +1597,19 @@ mod tests {
         );
 
+        let text = &to_utf16(text);
+        assert_eq!(
```
suggestion: would be nice to refactor this double-assert into a helper function

Manishearth reviewed (commented):
Overall looks good. Please fix anything marked issue/nit and we can merge this.

(Fine to force push again, now that I've reviewed the entire thing)

Thanks for doing this!

Manishearth commented on src/lib.rs:
```diff
@@ -208,120 +238,225 @@ impl<'text> InitialInfoExt<'text> {
+///
+/// Contains the text paragraphs and `BidiClass` of its characters.
+#[derive(PartialEq, Debug)]
+pub struct InitialInfoU16<'text> {
```
thought: it would be ideal for us to have a GenericBidiInfo, GenericInitialInfo, and GenericParagraph types that abstract over an encoding, where the UTF8 one is just a type alias.

This might be more work. I'm fine with having separate UTF16 types for now as long as they're in a module.

jfkthame replied:
Yes, that'd be nice; my grasp of Rust is still pretty weak, though, so I haven't attempted this as yet.

jfkthame commented:
Huh, I wanted to completely remove `pub` from the `TextSource` trait, and locally that works fine for me, but the CI build here doesn't seem to like it. Sigh.

jfkthame commented:
@Manishearth Thanks for all the review & comments. I think I've addressed all the nits/issues, just a couple of suggestions that I didn't do yet -- they seem fine in principle but my Rust is pretty weak still.

Apologies for the repeated force-push spam! It was all building fine for me locally, but took some repeated attempts to get the CI jobs happy. Not sure about the failing 1.36.0 job? Please let me know what I ought to do, if anything, or can we ignore it? Thanks!

mbrubeck commented:
> Not sure about the failing 1.36.0 job?

It looks like `sealed` 0.5.0 and its dependencies will not build on Rust 1.36.0.  We would need to increase our minimum supported Rust version to 1.56.0 to use this crate.

I tried downgrading to older versions of `sealed` but these failed to build on 1.36.0 for a different reason.  However, they might work with versions newer than 1.36.0 but older than 1.56.0

Manishearth commented:
We ought to be able to tweak our CI lockfile

mbrubeck commented:
> We ought to be able to tweak our CI lockfile

Unfortunately, `sealed` 0.5.0 requires `syn` 2.0.0 or higher, and there is no matching version of `syn` that compiles with Rust 1.36.  (`syn` uses Rust 2021 edition, which requires rustc 1.56.0 or later.)

`sealed` 0.4 and earlier use `syn` 1.x, but the `sealed` code itself will not build with rustc 1.36.0 in any published version. (It is missing `extern crate proc_macro;` which was required for using the `proc_macro` crate in Rust 2018 edition crates at the time.)

Manishearth commented:
Alright, in that case I'm fine dropping the CI requirement!

mbrubeck commented:
To be clear, that means increasing the MSRV to 1.56 and adding unconditional dependencies on `sealed` and `syn`.

As an alternative, what about implementing the [sealed trait pattern](https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed) “manually” (without a proc-macro crate)?

Manishearth commented:
I'm comfortable increasing to 1.56, but yes if we have an alternative we should use it

I didn't even realize we depended on syn, I would love for this to be a leaf crate

mbrubeck commented:
Currently we depend on `syn` only with the optional `serde` or `flamer` features enabled.  By default this crate has zero dependencies.

Manishearth commented:
ah, yeah, I'm fine with breaking msrv for those features only

we can tweak CI to not run those

jfkthame commented:
I took @mbrubeck's suggestion (thanks!) to implement the sealed-trait pattern manually, which seems straightforward, so that removes the dependency on the `sealed` crate.

Unfortunately, 1.36.0 still fails, because `REPLACEMENT_CHARACTER`, `from_u32()`, and `decode_utf16()` are all in the `std` version of `char`, not provided by the primitive `char` type.

@Manishearth Should we just provide local versions of these, too? They seem like they should be simple enough...

mbrubeck commented:
> Unfortunately, 1.36.0 still fails, because `REPLACEMENT_CHARACTER`, `from_u32()`, and `decode_utf16()` are all in the `std` version of `char`, not provided by the primitive `char` type.

You can use these as `core::char::REPLACEMENT_CHARACTER`, `core::char::from_u32`, etc.

jfkthame commented:
> > Unfortunately, 1.36.0 still fails, because `REPLACEMENT_CHARACTER`, `from_u32()`, and `decode_utf16()` are all in the `std` version of `char`, not provided by the primitive `char` type.
> 
> You can use these as `core::char::REPLACEMENT_CHARACTER`, `core::char::from_u32`, etc.

Huh, you're right -- the error message told me I needed `std`, but `core` also works. So it's sufficient to just say `use core::char;` at the beginning of the module to resolve this.

Next up is that I put a lot of the tests and expected-results into arrays, and then use `for t in tests { ... }` to run the same calls & assertions over a whole set of tests. That works fine for me with current Rust, but again, 1.36.0 doesn't like it, so looks like that'll need to be reworked. I'll try to have a look at it, but any hints/suggestions are welcome!

mbrubeck commented:
There are various options, but one simple one is to change the arrays to vecs: `let tests = vec![ /* ... */ ];`

It looks like Rust 1.36 also chokes on one use of `format!` to create an `&str` in the test module.  This can probably be replaced by a string literal with Unicode character escapes, or you can move the `format!` to a binding outside of the `tests` vector like:

```rust
        let fsi_rtl_pdi_ltr = format!("{}א{}a", chars::FSI, chars::PDI);
        let tests = vec![
            // ...
            (
                &fsi_rtl_pdi_ltr,
                // ...
```

Manishearth commented:
> Apologies for the repeated force-push spam! It was all building fine for me locally, but took some repeated attempts to get the CI jobs happy. Not sure about the failing 1.36.0 job? Please let me know what I ought to do, if anything, or can we ignore it? Thanks!

When dealing with a PR this big it's better to just make fixup commits and clean up the commit history later.

Manishearth commented:
Hmm, I went through all of the new interdiffs and could not see any of the issues being addressed. I might have missed an interdiff or a commit, I might have to go through this commit by commit again.

(This is why I was discouraging force pushes)

Manishearth commented on src/implicit.rs:
```diff
@@ -477,37 +481,30 @@ struct BracketPair {
+        let mut actual_index = level_run.start;
+
+        for ch in text.subrange(level_run.clone()).chars() {
+            let char_len = T::char_len(ch);
```
nit: let's use char_indices here instead of recomputing char_len. there's a chance the optimizer will figure this out, I'd rather not take that chance.

Again, the old code had the right structure: run char_indices and add to `level_run.start`.

jfkthame replied:
Done in [7866637](https://github.com/servo/unicode-bidi/pull/109/commits/786663784fc5f5c8b700c9b2fd7ebcb4227aeb04)

Manishearth commented on src/lib.rs:
```diff
@@ -697,76 +633,317 @@ impl<'text> BidiInfo<'text> {
+/// If doing so, you may prefer to use [`reordered_levels_per_char()`] instead
+/// to avoid non-character indices.
+///
+/// For an all-in-one reordering solution, consider using [`Self::reorder_visual()`].
```
nit: this line and the one above it is not relevant for a private API.

This should just be a low level explanation of the function: what it does (the first line of docs) and what spec stuff it covers (the mention of rule L1)

jfkthame replied:
Done in [69aba46](https://github.com/servo/unicode-bidi/pull/109/commits/69aba46d286f4e02a605b1b9612911358a3641c9)

Manishearth commented on src/lib.rs:
```diff
@@ -81,25 +81,85 @@ mod char_data;
 pub use crate::level::{Level, LTR_LEVEL, RTL_LEVEL};
 pub use crate::prepare::LevelRun;
+pub use crate::utf16::{
+    BidiInfo as BidiInfoU16, InitialInfo as InitialInfoU16, Paragraph as ParagraphU16,
```
issue: no, I want these types to not be in the root module. Please have `pub mod utf16` instead

jfkthame replied:
Done in [74a631e](https://github.com/servo/unicode-bidi/pull/109/commits/74a631e768b9c37197d78968f3b7df0faf4681ce)

Manishearth reviewed (changes requested):
Almost there, still a couple of my comments that haven't been completely addressed.

I didn't redo the commit by commit review but I went through my old comments and looked at what was fixed.

For future fixes please only make new commits. You can use `git commit --fixup <commit hash>` to stage a commit for being squashed into a previous commit later; we can clean up the commit history before merging.

Though the changes I have requested probably are fine in their own commits.

Manishearth commented on src/implicit.rs:
```diff
@@ -256,6 +256,7 @@ pub fn resolve_neutral<'a, D: BidiDataSource, T: TextSource<'a> + ?Sized>(
     original_classes: &[BidiClass],
     processing_classes: &mut [BidiClass],
 ) where
+    <T as TextSource<'a>>::CharIndexIter: Iterator<Item = (usize, char)>,
```
nit: all of these bounds should live on the TextSource trait, not on all these where clauses

Manishearth commented on src/lib.rs:
```diff
@@ -93,13 +94,74 @@ pub use crate::char_data::{bidi_class, HardcodedBidiData};
+/// This trait is sealed and cannot be implemented for types outside this crate.
+pub trait TextSource<'text>: private::Sealed
+where
+    <Self as TextSource<'text>>::CharIndexIter: Iterator<Item = (usize, char)>,
```
not: not in the where bounds, they can go on the associated types directly: `type CharIndexIterator: Iterator<...>`

Manishearth reviewed (approved):
looks good with one small change. feel free to force push / squash and I'll merge

Manishearth commented:
Thanks!
